Data/Primitive/MVar.hs

```diff
@@ -0,0 +1,133 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE MagicHash #-}
+{-# LANGUAGE UnboxedTuples #-}
+
+-- |
+-- Module      : Data.Primitive.ByteArray
+-- Copyright   : (c) Roman Leshchinskiy 2009-2012
+-- License     : BSD-style
+--
+-- Maintainer  : Roman Leshchinskiy <rl@cse.unsw.edu.au>
+-- Portability : non-portable
+--
+-- Primitive operations on @MVar@. This module provides a similar interface
+-- to "Control.Concurrent.MVar". However, the functions are generalized to
+-- work in any 'PrimMonad' instead of only working in 'IO'. Note that all
+-- of the functions here are completely deterministic. It is only the
+-- introduction of @forkIO@ that brings nondeterminism into the picture.
+-- Users who are interesting in using these functions in 'ST' will likely
+-- want to roll their own 'ST' variant of @forkIO@. Since such a function
+-- eschews that determinism expected by users of 'ST' (and users of
+-- 'primitive'), it is left to the user to implement it. 
+module Data.Primitive.MVar
+  ( MVar(..)
+  , newMVar
+  , isEmptyMVar
+  , newEmptyMVar
+  , putMVar
+#if __GLASGOW_HASKELL__ >= 708
+  , readMVar
+#endif
+  , takeMVar
+  , tryPutMVar
+#if __GLASGOW_HASKELL__ >= 708
+  , tryReadMVar
+#endif
+  , tryTakeMVar
+  ) where
+
+import Control.Monad.Primitive
+import Data.Primitive.Internal.Compat (isTrue#)
+import GHC.Exts (MVar#,newMVar#,takeMVar#,sameMVar#,putMVar#,tryTakeMVar#,
+  isEmptyMVar#,tryPutMVar#,(/=#))
+
+#if __GLASGOW_HASKELL__ >= 708
+import GHC.Exts (readMVar#,tryReadMVar#)
+#endif
+
+data MVar s a = MVar (MVar# s a)
+
+instance Eq (MVar s a) where
+  (MVar mvar1#) == (MVar mvar2#) = isTrue# (sameMVar# mvar1# mvar2#)
+
+-- | Create an 'MVar' which is initially empty.
```

Comment:
create a fresh mvar in the empty state.

"an" is for things that start with a vowel